ffpass/__init__.py

```diff
@@ -63,47 +62,61 @@ class WrongPassword(Exception):
     pass
 
 
+def decrypt_aes(decoded_item, master_password, global_salt):
+    entry_salt = decoded_item[0][0][1][0][1][0].asOctets()
+    iteration_count = int(decoded_item[0][0][1][0][1][1])
+    key_length = int(decoded_item[0][0][1][0][1][2])
+    assert key_length == 32
+
+    k = sha1(global_salt + master_password.encode('utf-8')).digest()
+    key = pbkdf2_hmac('sha256', k, entry_salt, iteration_count, dklen=key_length)
+
+    iv = b'\x04\x0e' + decoded_item[0][0][1][1][1].asOctets()
+    encrypted_value = decoded_item[0][1].asOctets()
+    cipher = AES.new(key, AES.MODE_CBC, iv)
+    return cipher.decrypt(encrypted_value)
+
+
 def getKey(directory: Path, masterPassword=""):
     dbfile: Path = directory / "key4.db"
+
     if not dbfile.exists():
         raise NoDatabase()
-    # firefox 58.0.2 / NSS 3.35 with key4.db in SQLite
+
     conn = sqlite3.connect(dbfile.as_posix())
     c = conn.cursor()
-    # first check password
-    c.execute("SELECT item1,item2 FROM metadata WHERE id = 'password';")
+    c.execute("""
+        SELECT item1,item2
+        FROM metadata
+        WHERE id = 'password';
+    """)
     row = next(c)
     globalSalt = row[0]  # item1
     item2 = row[1]
-    decodedItem2, _ = der_decode(item2)
-    entrySalt = decodedItem2[0][1][0].asOctets()
-    cipherT = decodedItem2[1].asOctets()
-    clearText = decrypt3DES(
-        globalSalt, masterPassword, entrySalt, cipherT
-    )  # usual Mozilla PBE
+
+    decodedItem2 = der_decode(item2)
+    clearText = decrypt_aes(decodedItem2, masterPassword, globalSalt)
+
     if clearText != b"password-check\x02\x02":
         raise WrongPassword()
     if args.verbose:
         print("password checked", file=sys.stderr)
+
     # decrypt 3des key to decrypt "logins.json" content
     c.execute("SELECT a11,a102 FROM nssPrivate;")
-    for row in c:
-        if row[1] == MAGIC1:
-            a11 = row[0]  # CKA_VALUE
-            break
-    else:
+    row = next(c)
+    a11 = row[0]   # CKA_VALUE
+    a102 = row[1]  # CKA_ID
+
+    if a102 != MAGIC1:
```

Comment:
I was thinking, it would probably be better to do a "real" SQL query: `c.execute("SELECT a11,a102 FROM nssPrivate WHERE a102 = ?", a102)`